Integration tests: give nodes a name and print that name in log file name

guggero opened the pull request:
Addresses #1134.
Now the log file names look like this:
```text
output-0-Bob-00000000.log
output-1-Alice-00000000.log
output-1-Alice-02d80659.log
output-2-Carol-00000000.log
output-3-Carol-00000000.log
output-4-Carol-00000000.log
output-5-Carol-00000000.log
output-6-Carol-00000000.log
output-7-Carol-00000000.log
output-8-Carol-00000000.log
output-9-Carol-00000000.log
...
```

I suspect that this might also solve #1133. I cannot say with certainty what the problem with termbin is, but I suspect that the files `output0.log` and `output1.log` just got too big and therefore couldn't be uploaded any more (maybe there is a size limit at termbin?).
Now, the output of node 0 and 1 are split into several files, so they remain smaller. But proper investigation is still necessary to not just postpone the problem.

Roasbeef commented on lntest/node.go:
```diff
@@ -262,7 +269,8 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
 	if *logOutput {
-		fileName := fmt.Sprintf("output%d.log", hn.NodeID)
+		fileName := fmt.Sprintf("output-%d-%s-%s.log", hn.NodeID,
+			hn.cfg.Name, hex.EncodeToString(hn.PubKey[:logPubKeyBytes]))
```
The node's public key hasn't yet been instantiated at this point. If you look at the travis run, you'll see it's all zeroes for most of the nodes. I think adding the public key is definitely a good idea, but it may need to be done further down the line, as we actually don't know the public key of the node until it starts...

guggero replied:
Ah, I was wondering about that. I thought that most IDs were all zeroes because the integration tests were set up this way (since some IDs were initialized).
Didn't occur to me it just might be a timing problem.

I'll look into it.

guggero commented:
I found a solution that seems to work and isn't too complicated in my opinion:

Until we know the node's PubKey, we create a temporary file name. After the node has done its work, we simply rename the log file and include the correct PubKey.

halseth commented on lntest/node.go:
```diff
@@ -307,11 +350,18 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
 		hn.cmd.Process.Kill()
+
+		// Make sure log file is closed and renamed if necessary.
+		hn.finalizeLogfile()
```
Since we kill the process, I think maybe this will be called from the goroutine regardless?

halseth commented on lntest/node.go:
```diff
@@ -266,12 +271,50 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
+				hn.cfg.Name, hex.EncodeToString(randomBytes))
+
+			// Once the node has done its work, the log file can be renamed.
+			hn.finalizeLogfile = func() {
```
since this method is used only inseide start, it shouldn't need to be part of the node struct.

halseth commented on lntest/node.go:
```diff
@@ -266,12 +271,50 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
+						hex.EncodeToString(hn.PubKey[:logPubKeyBytes]))
+					err := os.Rename(fileName, newFileName)
+					if err != nil {
+						lndError <- errors.Errorf(
```
Don't think we can assume that there will be someone to read from this channel at this point. Better to just print the error.

halseth commented on lntest/node.go:
```diff
@@ -266,12 +271,50 @@ func (hn *HarnessNode) start(lndError chan<- error) error {
+			rand.Seed(time.Now().UnixNano())
+			randomBytes := make([]byte, 4)
+			rand.Read(randomBytes)
+			fileName = fmt.Sprintf("output-%d-%s-tmp__%s.log", hn.NodeID,
```
The node IDs are unique, making sure the file names won't collide?

guggero commented:
I was working with the (incorrect) assumption that a node with the same NodeID could be restarted and possibly get another PubKey. That would have resulted in multiple log files per NodeID, therefore the random number part.
But apparently this is not necessary, so I removed the part with the random bytes.

Thank you for your input, @halseth. `cmd.wait()` does indeed return if the process is killed so the additional calls to `finalizeLogfile()` were not necessary.
I addressed the other items as well and updated the commit.
